### PR TITLE
work on moving theming to plugins

### LINKED ISSUE #5600: finish implementing plugin-based theming mechanism
https://www.notion.so/learningequality/Parameterized-Kolibri-themes-MVP-fc38e9f215924ed1abf42e1b063b9333

### LINKED ISSUE #5617: upgrade perseus
### Observed behavior

need to update translations and bump to latest version

### Expected behavior

…

### User-facing consequences

…

### Errors and logs

…

### Steps to reproduce

…

### Context

…

### PR DESCRIPTION
**Pull request by indirectlylit**
### Summary

* Puts themes in a default theme plugin
* Allows background image to be set either by theme or management command
* Allows logos in side nav, sign-in screen, and app bar to be customized

### Reviewer guidance

This provides some basic customization parameters

### References

* fixes #5600
* fixes #5617

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by nucleogenesis** [comment]
@indirectlylit I checked this out but I cannot access the site at all once I check out this branch. I get a 500 from an AssertionError on the server.

```
ERROR    Internal Server Error: /en/user/
Traceback (most recent call last):
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/core/handlers/base.py", line 217, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/core/handlers/base.py", line 215, in _get_response
    response = response.render()
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/response.py", line 107, in render
    self.content = self.rendered_content
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/response.py", line 84, in rendered_content
    content = template.render(context, self._request)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/backends/django.py", line 66, in render
    return self.template.render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 207, in render
    return self._render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 199, in _render
    return self.nodelist.render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 990, in render
    bit = node.render_annotated(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 957, in render_annotated
    return self.render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/loader_tags.py", line 177, in render
    return compiled_parent._render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 199, in _render
    return self.nodelist.render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 990, in render
    bit = node.render_annotated(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/base.py", line 957, in render_annotated
    return self.render(context)
  File "/home/jacob/.local/share/virtualenvs/kolibri--v441dVg/lib/python3.5/site-packages/django/template/library.py", line 203, in render
    output = self.func(*resolved_args, **resolved_kwargs)
  File "/home/jacob/Code/LearningEquality/kolibri/kolibri/core/templatetags/kolibri_tags.py", line 136, in kolibri_theme
    return mark_safe(template.format(theme=json.dumps(ThemeHook().theme)))
  File "/home/jacob/Code/LearningEquality/kolibri/kolibri/plugins/hooks.py", line 213, in inner
    assert len(instance.registered_hooks) == 1
AssertionError
ERROR    "GET /en/user/ HTTP/1.1" 500 179080
```

First  I ran `pip install -e .` 

Then - I set an image `kolibri manage background set -h /path/to/image`. Then I got the above error.

Then I ran `kolibri manage background unset` and still got the error.

After checking out the release branch and successfully starting the server, I then re-checked out this branch and got the error again.

File: kolibri/core/hooks.py [hunk 1]
```diff
@@ -62,49 +63,52 @@ def plugin_url(self, plugin_class, url_name):
...
     """
 
-    class Meta:
-
-        abstract = True
+    def validateBrandColors(self, theme):
+        if "brandColors" not in theme:
+            logger.error("brand colors not defined by theme")
+            return False
+        required_colors = ["primary", "secondary"]
```
Comment (rtibbles) [inline kolibri/core/hooks.py]: Might be good to define these constants in module scope for testing purposes?

File: kolibri/core/hooks.py [hunk 2]
```diff
@@ -62,49 +63,52 @@ def plugin_url(self, plugin_class, url_name):
...
-                theme[key] = getattr(hook, key, theme[key])
-            once = True
-
-        return theme or default
+        theme = self.registered_hooks[0].theme
+
+        self.validateBrandColors(theme)
+
+        # if a background image has been locally set using the `manage background` command, use it
+        if os.path.exists(os.path.join(settings.MEDIA_ROOT, "background.jpg")):
```
Comment (rtibbles) [inline kolibri/core/hooks.py]: Works for now - do we not allow pngs?

Might be good to store the filename somewhere instead.

File: kolibri/core/webpack/hooks.py [hunk 3]
```diff
@@ -521,7 +521,7 @@ class FrontEndCoreHook(WebpackInclusionHook):
 
     def __init__(self, *args, **kwargs):
         super(FrontEndCoreHook, self).__init__(*args, **kwargs)
-        assert len(list(self.registered_hooks)) <= 1, "Only one core asset allowed"
```
Comment (rtibbles) [inline kolibri/core/webpack/hooks.py]: `self.registered_hooks` is a generator, so it doesn't have a length, hence it being coerced to a list first.

File: kolibri/plugins/hooks.py [hunk 4]
```diff
@@ -203,6 +203,19 @@ def inner(instance, *args, **kwargs):
...
 
 
+def only_one_registered(func):
+    """
+    Ensures that only one hook of this type is registered at a time
+    """
+
+    @functools.wraps(func)
+    def inner(instance, *args, **kwargs):
+        assert len(instance.registered_hooks) == 1
```
Comment (rtibbles) [inline kolibri/plugins/hooks.py]: As above, need to coerce to something that has a length.

If taking the len of a list is unappealing can do something like:
`sum(1 for i in instance.registered_hooks)` to get the length.

File: kolibri/plugins/hooks.py [hunk 5]
```diff
@@ -337,8 +350,7 @@ def registered_hooks(self):
         hook is accessed, it's also instantiated, giving it a chance to re-do
         things that it wants done on every event.
         """
-        for HookClass in self._meta.registered_hooks:
-            yield HookClass()
+        return [HookClass() for HookClass in self._meta.registered_hooks]
```
Comment (rtibbles) [inline kolibri/plugins/hooks.py]: Would rather not change this behaviour unless there is a very good reason to prefer a list over a generator (as it touches very disparate parts of the backend code and could cause unintended consequences).

File: kolibri/plugins/user/assets/src/views/SignInPage/index.vue [hunk 6]
```diff
@@ -281,10 +281,15 @@
         return urls['kolibri:core:guest']();
       },
       backgroundImageStyle() {
-        return {
-          backgroundColor: this.$themeTokens.primary,
-          backgroundImage: `linear-gradient(rgba(0, 0, 0, 0.7), rgba(0, 0, 0, 0.7)), url(${require('./background.jpg')})`,
-        };
+        if (global.kolibriTheme.signInBackground) {
```
Comment (rtibbles) [inline kolibri/plugins/user/assets/src/views/SignInPage/index.vue]: Would prefer to consolidate all references to the global API in the theme module, then reference that through imports where it ends up being used - just to decouple our backend injection from frontend usage.

File: kolibri/plugins/user/management/commands/background.py [hunk 7]
```diff
@@ -56,23 +39,27 @@ def handle(self, *args, **options):
                 )
             )
 
-            new_img_path = os.path.abspath(os.path.expanduser(options["destination"]))
-            if not os.path.exists(new_img_path):
+            new_img = os.path.abspath(os.path.expanduser(options["new-image"]))
```
Comment (rtibbles) [inline kolibri/plugins/user/management/commands/background.py:44]: I know this is existing functionality, but looks like we don't validate the file extension/mimetype?

File: kolibri/deployment/default/settings/base.py [hunk 8]
```diff
@@ -240,6 +240,8 @@
 
 STATIC_URL = urljoin(path_prefix, "static/")
 STATIC_ROOT = os.path.join(conf.KOLIBRI_HOME, "static")
+MEDIA_URL = urljoin(path_prefix, "media/")
```
Comment (rtibbles) [inline kolibri/deployment/default/settings/base.py:242]: :+1:

**Review by rtibbles — DISMISSED** [review]
Several things that I have strong opinions about, but am happy to change my mind.

File: kolibri/plugins/user/management/commands/background.py (hunk 7 above)
Comment (indirectlylit) [reply kolibri/plugins/user/management/commands/background.py:44]: this should work with png, svg, jpg, gif, etc... I had intentionally left it permissive

**Comment by indirectlylit** [comment]
ready for re-review and possibly merge

File: kolibri/core/assets/src/views/ColourPicker.vue [hunk 9]
```diff
@@ -1,139 +0,0 @@
...
-      </div>
-    </div>
-  </div>
-
-</template>
-
-
-<script>
-
-  import { Chrome } from 'vue-color';
```
Comment (rtibbles) [inline kolibri/core/assets/src/views/ColourPicker.vue:43]: Can remove this package from package.json.

File: kolibri/core/theme_hook.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    class Meta:
+        abstract = True
+
+    @property
+    @hooks.only_one_registered
+    def cacheKey(self):
+        theme = list(self.registered_hooks)[0].theme
+        return parse.quote(
+            "{}-{}-{}".format(
+                kolibri.__version__, theme[THEME_NAME], theme[THEME_VERSION]
```
Comment (rtibbles) [inline kolibri/core/theme_hook.py:128]: As we discussed, the only disadvantage is that this won't currently expire for the background change management command, but I can live with that needing a restart.

File: kolibri/core/theme_hook.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+TOP_LOGO = "topLogo"
+IMG_SRC = "src"
+IMG_STYLE = "style"
+IMG_ALT = "alt"
+SHOW_K_FOOTER_LOGO = "showKolibriFooterLogo"
+
+# This is the image file name that will be used when customizing the sign-in background
+# image using the 'kolibri manage background' command. It does not attempt to use a file
+# extension (like .jpg) because we don't know if it's a JPG, SVG, PNG, etc...
+DEFAULT_BG_IMAGE_FILE = "background_image"
```
Comment (rtibbles) [inline kolibri/core/theme_hook.py:49]: I assume most browsers are smart enough to not need the file extension for the mimetype hinting.

File: kolibri/plugins/default_theme/kolibri_plugin.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+from __future__ import unicode_literals
+
+from django.contrib.staticfiles.templatetags.staticfiles import static
+
+from kolibri.plugins.base import KolibriPluginBase
+
+from kolibri.core import theme_hook
+
+
+class KolibriThemeTest(KolibriPluginBase):
```
Comment (rtibbles) [inline kolibri/plugins/default_theme/kolibri_plugin.py]: Maybe not the final name you intended?

**Review by rtibbles — COMMENTED** [review]
Nothing blocking, codewise, but some possible cleanup.

File: kolibri/core/theme_hook.py (hunk 11 above)
Comment (indirectlylit) [reply kolibri/core/theme_hook.py:49]: yes, I believe so

File: kolibri/core/theme_hook.py (hunk 10 above)
Comment (indirectlylit) [reply kolibri/core/theme_hook.py:128]: yes, I think for now there's no strong use case

**Review by rtibbles — APPROVED** [review]

**Comment by radinamatic** [comment]
Testing the asset from here on Ubuntu 16.4, not seeing anything that shouldn't be there :+1:

**Pull request merged**